File: problems/2517_maximum-tastiness-of-candy-basket/solution.cpp
class Solution {
  public:
    int maximumTastiness(vector<int> &price, int k) {
        vector<int> p(price);
        sort(p.begin(), p.end());
        auto feasible = [&](int x) {
            int count = 1;
            int last = p[0];
            for (size_t i = 1; i < p.size(); i++) {
                if (p[i] - last >= x) {
                    count++;
                    last = p[i];
                }
            }
            return count >= k;
        };
        int lo = 0, hi = p.back() - p[0];
        while (lo < hi) {
            int mid = lo + (hi - lo + 1) / 2;
            if (feasible(mid))
                lo = mid;
            else
                hi = mid - 1;
        }
        return lo;
    }
};

File: problems/2517_maximum-tastiness-of-candy-basket/solution.go
import "sort"

func maximumTastiness(price []int, k int) int {
	p := append([]int(nil), price...)
	sort.Ints(p)
	feasible := func(x int) bool {
		count := 1
		last := p[0]
		for _, v := range p[1:] {
			if v-last >= x {
				count++
				last = v
			}
		}
		return count >= k
	}
	lo, hi := 0, p[len(p)-1]-p[0]
	for lo < hi {
		mid := lo + (hi-lo+1)/2
		if feasible(mid) {
			lo = mid
		} else {
			hi = mid - 1
		}
	}
	return lo
}
